specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,743 @@
+
+#### SSZ encoding strategy (with or without Snappy)
+
+The [SimpleSerialize (SSZ) specification](https://github.com/ethereum/eth2.0-specs/blob/192442be51a8a6907d6401dffbf5c73cb220b760/specs/simple-serialize.md) outlines how objects are SSZ-encoded. If the Snappy variant is selected, we feed the serialised form to the Snappy compressor on encoding. The inverse happens on decoding.
+
+**Encoding-dependent header:** Req/Resp protocols using the `ssz` or `ssz_snappy` encoding strategies MUST prefix all encoded and compressed (if applicable) payloads with an unsigned [protobuf varint](https://developers.google.com/protocol-buffers/docs/encoding#varints).
+
+Note that parameters defined as `[]VariableName` are SSZ-encoded containerless vectors.
+
+### Messages
+
+#### Hello
+
+**Protocol ID:** ``/eth2/beacon_chain/req/hello/1/``
+
+**Content**:
+```
+(
+  fork_version: bytes4
+  finalized_root: bytes32
+  finalized_epoch: uint64
+  head_root: bytes32
+  head_slot: uint64
+)
+```
+The fields are:
+
+- `fork_version`: The beacon_state `Fork` version
+- `finalized_root`: The latest finalized root the node knows about
+- `finalized_epoch`: The latest finalized epoch the node knows about
+- `head_root`: The block hash tree root corresponding to the head of the chain as seen by the sending node
+- `head_slot`: The slot corresponding to the `head_root`.
+
+Clients exchange hello messages upon connection, forming a two-phase handshake. The first message the initiating client sends MUST be the hello message. In response, the receiving client MUST respond with its own hello message.
+
+Clients SHOULD immediately disconnect from one another following the handshake above under the following conditions:
+
+1. If `fork_version` doesn’t match the local fork version, since the client’s chain is on another fork. `fork_version` can also be used to segregate testnets.
+2. If the (`finalized_root`, `finalized_epoch`) shared by the peer is not in the client's chain at the expected epoch. For example, if Peer 1 sends (root, epoch) of (A, 5) and Peer 2 sends (B, 3) but Peer 1 has root C at epoch 3, then Peer 1 would disconnect because it knows that their chains are irreparably disjoint.
+
+Once the handshake completes, the client with the lower `finalized_epoch` or `head_slot` (if the clients have equal `finalized_epoch`s) SHOULD request beacon blocks from its counterparty via the `BeaconBlocks` request.
+
+#### Goodbye
+
+**Protocol ID:** ``/eth2/beacon_chain/req/goodbye/1/``
+
+**Content:**
+```
+(
+  reason: uint64
+)
+```
+Client MAY send goodbye messages upon disconnection. The reason field MAY be one of the following values:
+
+- 1: Client shut down.
+- 2: Irrelevant network.
+- 3: Fault/error.
+
+Clients MAY use reason codes above `128` to indicate alternative, erroneous request-specific responses.
+
+The range `[4, 127]` is RESERVED for future usage.
+
+#### BeaconBlocks
```

Comment:
I think we need to add one for `BeaconState` as well, to request the last finalized state to start syncing

Reply:
for interop, we start with full sync clients from genesis - syncing state should probably be landed together with light clients and a strategy for deciding _which_ states should be available for download (all, certain epochs, one per month, etc)